server/src/main/java/org/elasticsearch/index/mapper/ObjectMapper.java

```diff
@@ -533,15 +537,22 @@ private static Map<String, Mapper> buildMergedMappers(
             MergeReason reason,
             MapperMergeContext objectMergeContext
         ) {
-            Map<String, Mapper> mergedMappers = null;
+            Map<String, Mapper> mergedMappers = new HashMap<>(existing.mappers);
             for (Mapper mergeWithMapper : mergeWith) {
-                Mapper mergeIntoMapper = (mergedMappers == null ? existing.mappers : mergedMappers).get(mergeWithMapper.simpleName());
+                Mapper mergeIntoMapper = mergedMappers.get(mergeWithMapper.simpleName());
 
-                Mapper merged;
                 if (mergeIntoMapper == null) {
-                    merged = mergeWithMapper;
+                    if (mergeWithMapper instanceof ObjectMapper om) {
+                        ObjectMapper withoutMappers = om.withoutMappers();
+                        boolean added = objectMergeContext.addFieldIfPossible(mergedMappers, withoutMappers);
+                        if (added) {
+                            mergedMappers.put(om.simpleName(), withoutMappers.merge(mergeWithMapper, reason, objectMergeContext));
```

Comment:
are we not putting twice now? Once within addFieldIfPossible and once here ? The value is different but the key is the same?

Reply:
Hopefully, `MergeResult#addNewObjectMapper` also makes this a little clearer now. But It still essentially puts the same key twice. We could prevent that by exposing a method that decrements the field budget but doesn't add the Mapper to the `mergedMappers` map. However, I'd prefer not exposing such a method so that the process of checking the budget, decrementing it, and executing a consumer that puts the mapper to the `mergedMappers` map is encapsulated and callers can't forget a step.

If you feel strongly that this is not a good idea or not readable, we could also expose `MapperMergeContext.NewFieldsBudget#decrementIfPossible` and make the caller responsible for putting the mapper to the `mergedMappers` map.